### PR TITLE
Implement core support for `| resize(width, height, options)` filter

### LINKED ISSUE #5059: Add core support for the | resize(width, height, options) filter
The `| resize(width, height, options)` filter is one of the most included third party add-ons available for October, with the following plugins providing variations of it to over 23,000 installations: 

- https://octobercms.com/plugin/toughdeveloper-imageresizer
- https://octobercms.com/plugin/abwebdevelopers-imageresize
- https://octobercms.com/plugin/raideer-thumbnail
- https://octobercms.com/plugin/aspendigital-resizer
- https://octobercms.com/plugin/code200-imageking
- https://octobercms.com/plugin/popcornphp-imagecompress

More information to come, issue / design statement is WIP

### PR DESCRIPTION
**Pull request by LukeTowers**
Fixes #5059.

Example post processing plugin here: https://github.com/LukeTowers/oc-tinypng-plugin

### Related PRs:
- https://github.com/octobercms/docs/pull/472

### Replaces:
- https://octobercms.com/plugin/abwebdevelopers-imageresize (main functionality, support for Intervention Image can be retained by that plugin and provided as an extension to this functionality)
- https://octobercms.com/plugin/toughdeveloper-imageresizer (TinyPNG functionality provided by https://octobercms.com/plugin/luketowers-tinypng) 
- https://octobercms.com/plugin/raideer-thumbnail
- https://octobercms.com/plugin/aspendigital-resizer

### Remaining work:
- [x] Implement `| imageHeight`, `| imageWidth` filters
- [x] Implement support for the image resizing events when resizing File models
- [x] Cleanup inline documentation for ImageResizer class
- [ ] Complete unit tests
- [x] Add support for a `type: image` column type

### Unit tests to be completed:
- [x] Relative URLs for the four sources work (theme, plugin, media library, system file model)
- [x] Absolute URLs for the four sources work (theme, plugin, media library, system file model)
- [ ] Accessing a file that should not be accessed (i.e something in storage/app/uploads/protected or another custom storage directory not explicitly allowed in the getAvailableSources() method) is denied
- [ ] Generated resize images can't have their paths manipulated to break out of their intended target locations
- [ ] Valid resizer URL gets correctly signed and validated
- [x] Invalid resizer URL (target URL doesn't match signed identifier) fails validation
- [x] `| imageWidth` and `| imageHeight` filters return the correct values
- [x] `| resize()` filter returns the original string if unable to create a valid image from it
- [x] Falsely values provided as width or height to the resizer result in the value of `0` internally
- [ ] Images in a subfolder with a space character in the folder name in the media library are identified correctly (see https://github.com/abwebdevelopers/oc-imageresize-plugin/issues/30)
- [ ] Resizer can correctly find images when October is installed in a subfolder (see https://github.com/abwebdevelopers/oc-imageresize-plugin/issues/28)

### CODE REVIEW

**Comment by mjauvin** [comment]
@LukeTowers so this is improving the getThumb() in that it can work on an image from the media library in ADDITION TO an image object ?

**Comment by LukeTowers** [comment]
Sort of, this will rewrite the getThumb logic to be located in a more generic place and support handling File models, Media library items, and hopefully any other asset that the application can access (plugin / theme assets, etc)

File: modules/system/classes/SystemController.php [hunk 1]
```diff
@@ -35,9 +36,34 @@ public function combine($name)
...
+     * @param string $identifier The identifier used to retrieve the image configuration
+     * @param string $encodedUrl The double-encoded URL of the resized image, see https://github.com/octobercms/october/issues/3592#issuecomment-671017380
+     * @return RedirectResponse
+     */
+    public function resizer(string $identifier, string $encodedUrl)
+    {
+        $resizedUrl = ImageResizer::getValidResizedUrl($identifier, $encodedUrl);
+        if (empty($resizedUrl)) {
+            return response('Invalid identifier or redirect URL', 400);
+        }
```
Comment (LukeTowers) [inline modules/system/classes/SystemController.php:57]: @bennothommo could we get a unit test for both of these code paths (correctly signed URL being validated vs incorrectly signed URL being rejected)?

**Comment by bennothommo** [comment]
Awesome work with this, @LukeTowers!

**Comment by LukeTowers** [comment]
Thanks @bennothommo, I'm looking forward to getting this into production!

**Comment by bennothommo** [comment]
@LukeTowers just checking - is this supposed to support direct paths to images as well, as opposed to URLs or FileModel instances?

**Comment by LukeTowers** [comment]
Yes? If you can come up with a valid reason for a developer to pass an input that should work to the resizer then it should support it.

File: modules/system/classes/ImageResizer.php [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+use Exception;
+use SystemException;
+use File as FileHelper;
+use System\Models\File as SystemFileModel;
+use October\Rain\Database\Attach\File as FileModel;
+use October\Rain\Database\Attach\Resizer as DefaultResizer;
+
+/**
+ * $width = numeric, 'auto' | false | null
+ * $height = numeric, 'auto' | false | null
```
Comment (bennothommo) [inline modules/system/classes/ImageResizer.php]: @LukeTowers your docs here state that width and height can be specified as a string `auto`, however, I don't see any handling of that value. The docblock for the constructor says only integer, bool and null are accepted for those, and they are converted to an int on construction.

Comment (LukeTowers) [reply modules/system/classes/ImageResizer.php]: @bennothommo oversight on my part, just need to make sure that the constructor converts 'auto' to 0.

Comment (bennothommo) [reply modules/system/classes/ImageResizer.php]: @LukeTowers no worries - I've added that in for you.

File: tests/unit/system/classes/ImageResizerTest.php [hunk 3]
```diff
@@ -4,68 +4,187 @@
...
         $providedPath = '/plugins/october/demo/assets/NOTPRESENT.png';
     }
+
+    protected function setUpStorage()
+    {
+        $this->app->useStoragePath(base_path('storage/temp'));
+
+        config(['filesystems.disks.test_local' => [
+            'driver' => 'local',
+            'root'   => storage_path('app'),
```
Comment (LukeTowers) [inline tests/unit/system/classes/ImageResizerTest.php:329]: Shouldn't this be `storage_path('temp/app');`

Comment (bennothommo) [reply tests/unit/system/classes/ImageResizerTest.php:329]: @LukeTowers nope - the call above sets the storage path to `temp` already - see `$this->app->useStoragePath` :).

**Comment by bennothommo** [comment]
@LukeTowers unless I'm mistaken, it appears the new events provided in this functionality won't fire for file models, so plugins extending this functionality wouldn't be able apply optimisations to uploaded files. Should these events be extended to those types of sources too?

EDIT: I mean the `system.resizer.processResize` and `system.resizer.afterResize` events.

**Comment by mjauvin** [comment]
@LukeTowers I never got a response to my comment about the beforeResize event being documented but not fired anywhere...

**Comment by LukeTowers** [comment]
@mjauvin sorry, didn't see that. After thinking about it I couldn't find a legitimate use case for it so I didn't implement it. Still need to clean up the inline docs.

File: modules/system/classes/ImageResizer.php [hunk 4]
```diff
@@ -624,6 +624,21 @@ public static function normalizeImage($image)
...
+     * Normalize the provided path to Unix style directory seperators to ensure
+     * that path manipulation operations succeed regardless of environment
+     *
+     * NOTE: Can't use October\Rain\FileSystem\PathResolver because it prepends
+     * the current working directory to relative paths
+     *
+     * @param string $path
+     * @return string
+     */
+    protected static function normalizePath($path)
```
Comment (LukeTowers) [inline modules/system/classes/ImageResizer.php:642]: @bennothommo please take note, I believe the common spelling of normalize in tech is `normalize`, I noticed when working on this that the recent PathResolver class uses `normalise`. If you could fix that before we release 469 I'd appreciate it :)

Comment (bennothommo) [reply modules/system/classes/ImageResizer.php:642]: Sorry @LukeTowers - only just saw this. Fixed: https://github.com/octobercms/library/commit/e2ed5fbdbc20e3761a491889817fcb7e7e001532

**Comment by PubliAlex** [comment]
Hello,

Is there some documentation somewhere about this new core filter ? I looked for some doc on the octobercms website without success. Thank you

**Comment by LukeTowers** [comment]
Yes, it'll be added to the marketplace site when 1.1 is available on the marketplace, until then use the RainLab.Docs plugin or look at the docs repo directly: https://github.com/octobercms/docs/blob/master/services-image-resizing.md, https://github.com/octobercms/docs/blob/master/markup-filter-resize.md, https://github.com/octobercms/docs/blob/master/markup-filter-image-width.md, https://github.com/octobercms/docs/blob/master/markup-filter-image-height.md

**Pull request merged**